src/silentpayments.rs

```diff
@@ -0,0 +1,546 @@
+            let silentpayments_public_data = SilentpaymentsPublicData(silentpayments_public_data);
+            Ok(silentpayments_public_data)
+        } else {
+            Err(SilentpaymentsPublicDataError::ParseFailure)
+        }
+    }
+
+    /// Create Silent Payment shared secret.
+    pub fn recipient_create_shared_secret<C: Verification>(&self, secp: &Secp256k1<C>, recipient_scan_key: &SecretKey) -> Result<[u8; 33], SilentpaymentsPublicDataError> {
+        let mut output33 = [0u8; 33];
+
+        let res = unsafe {
+            secp256k1_silentpayments_recipient_create_shared_secret(
+                secp.ctx().as_ptr(),
+                output33.as_mut_c_ptr(),
+                recipient_scan_key.as_c_ptr(),
+                self.as_c_ptr(),
+            )
+        };
+
+        if res == 1 {
+            Ok(output33)
+        } else {
+            Err(SilentpaymentsPublicDataError::SharedSecretFailure)
+        }
+    }
+}
+
+/// Found outputs struct
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub struct SilentpaymentsFoundOutput(ffi::SilentpaymentsFoundOutput);
+
+impl CPtr for SilentpaymentsFoundOutput {
+    type Target = ffi::SilentpaymentsFoundOutput;
+
+    /// Obtains a const pointer suitable for use with FFI functions.
+    fn as_c_ptr(&self) -> *const Self::Target { &self.0 }
+
+    /// Obtains a mutable pointer suitable for use with FFI functions.
+    fn as_mut_c_ptr(&mut self) -> *mut Self::Target { &mut self.0 }
+}
+
+impl fmt::Display for SilentpaymentsFoundOutput {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+
+        let pubkey = XOnlyPublicKey::from(self.0.output.clone());
+
+        let buffer_str = pubkey.serialize()
+            .iter()
+            .map(|byte| format!("{:02x}", byte))
+            .collect::<String>();
+
+        write!(f, "{}", buffer_str)
+    }
+}
+
+impl SilentpaymentsFoundOutput {
+    /// Create a new `SilentpaymentsFoundOutput` object from a ffi::SilentpaymentsFoundOutput.
+    pub fn empty() -> SilentpaymentsFoundOutput {
+        SilentpaymentsFoundOutput(ffi::SilentpaymentsFoundOutput::empty())
+    }
+}
+
+/// Output scan errors
+#[derive(Debug, Clone, Copy, Eq, PartialEq, PartialOrd, Ord, Hash)]
+pub enum OutputScanError {
+    /// Unexpected failures
+    Failure,
+}
+
+#[cfg(feature = "std")]
+impl std::error::Error for OutputScanError {}
+
+impl fmt::Display for OutputScanError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> Result<(), fmt::Error> {
+        match self {
+            OutputScanError::Failure => write!(f, "Failed to scan outputs"),
+        }
+    }
+}
+
+/// Scan for Silent Payment transaction outputs.
+pub fn silentpayments_recipient_scan_outputs<C: Verification, L>(
+    secp: &Secp256k1<C>,
+    tx_outputs: &[&XOnlyPublicKey],
+    recipient_scan_key: &SecretKey,
+    public_data: &SilentpaymentsPublicData,
+    recipient_spend_pubkey: &PublicKey,
+    label_lookup: SilentpaymentsLabelLookupFunction,
+    label_context: L,
```

Comment:
This requires users to use unsafe, and is maybe even unsound to accept a function that can return pointers to a "safe" (not annotated unsafe) function.

Reply:
This is also unsound on Rust < 1.81, see https://blog.rust-lang.org/2024/09/05/Rust-1.81.0.html#abort-on-uncaught-panics-in-extern-c-functions. To fix that you have to `catch_unwind` and `abort` manually when using older versions.

Reply:
I think I managed to create a safe abstraction for this callback (given the current restrictions, which might change as https://github.com/bitcoin-core/secp256k1/pull/1519 is not merged yet). I added it (2b3f633) to [a branch on my fork](https://github.com/antonilol/rust-secp256k1/tree/bip352-recipient-scan-outputs-safe-callback), feel free to add (may be modified) it here.

Reply:
Thanks @antonilol for all your contributions, here and to other projects like electrs! They are very helpful.

Reply:
Thanks for working on this and proposing a better approach.
But I haven't added the commit yet because it still has some TODO and the example file is not using the modified function.
I would like to better understand your proposal and how we can implement the TODOs and call the function from the example file.

Reply:
First TODO: the C library allows you to pass a null pointer for the label lookup function, but the function signature on the Rust side does not, [this type definition](https://github.com/jlest01/rust-secp256k1/blob/bip352-silentpayments-module/secp256k1-sys/src/lib.rs#L1283) must be wrapped in `Option<...>` to allow null pointers.
Second TODO: the C library checks that the context pointer is non null when the function pointer is non null, in my opinion the C library should treat it as opaque and 100% managed by the user so should not check it, just pass it to the callback ([comment thread](https://github.com/bitcoin-core/secp256k1/pull/1519#discussion_r1748595895), same link as in the code). Note this will not be an issue for the Rust side anymore when TODO 1 is fixed, but at first I thought that could be harder because bindings are normally generated.

Function pointer + context pointer is a C thing, in Rust a closure automatically captures the variables it needs. To allow users to pass closures, the closure needs to be "converted" to a function pointer and context pointer for the C side to use.
This safe wrapper stores the state part of the closure (a Rust closure can be seen as a combination of a function and state) on the stack and passes the pointer to that as the context pointer. In the callback the function and the state are but "back together" and called.
To add to this, the C side expects a pointer to the return value. The C function uses this after the callback returned (obviously), so this pointer needs to live for longer than the callback, but as long as the safe wrapper is enough, so that return value is stored on the stack as well. This storage (the variable named `storage`) is passed with the closure state to the C side as the context pointer.

Calling from the example (or from any user code) should be straightforward, these should (roughly) be the changes needed to the example:
- make the cache a Vec<LabelCacheEntry> for simplicity
- change the callback function (in the argument list) to
```rust
|label| cache.iter().find(|cache_entry| cache_entry.label == label).map(|cache_entry| cache_entry.label_tweak)
```
`cache` will be a captured variable and the safe wrapper will take care of getting it to the C side.